File: lib/std/math.lua
-- Standard library math module for Sapphire (Lua target)

local math_mod = {}

function math_mod.abs(x)
  if x < 0 then
    return -x
  end
  return x
end

function math_mod.sqrt(x)
  if x < 0 then
    return nil
  end
  return math.sqrt(x)
end

function math_mod.min(a, b)
  if a < b then
    return a
  end
  return b
end

function math_mod.max(a, b)
  if a > b then
    return a
  end
  return b
end

function math_mod.safe_div(a, b)
  if b == 0 then
    return nil
  end
  -- TODO: Validate against Lua 5.1 / LuaJIT.
  if math.type and math.type(a) == "integer" and math.type(b) == "integer" then
    return math.floor(a / b)
  end
  return a / b
end

function math_mod.log(x, base)
  if x <= 0 then
    return nil
  end
  if base ~= nil then
    if base <= 0 or base == 1 then
      return nil
    end
    return math.log(x) / math.log(base)
  end
  return math.log(x)
end

function math_mod.pow(base, exp)
  return base ^ exp
end

function math_mod.ceil(x)
  return math.ceil(x)
end

function math_mod.floor(x)
  return math.floor(x)
end

-- Generic monomorphization aliases
math_mod.abs__int = math_mod.abs
math_mod.abs__float = math_mod.abs
math_mod.sqrt__int = math_mod.sqrt
math_mod.sqrt__float = math_mod.sqrt
math_mod.min__int = math_mod.min
math_mod.min__float = math_mod.min
math_mod.max__int = math_mod.max
math_mod.max__float = math_mod.max
math_mod.safe_div__int = math_mod.safe_div
math_mod.safe_div__float = math_mod.safe_div
math_mod.log__int = math_mod.log
math_mod.log__float = math_mod.log
math_mod.pow__int = math_mod.pow
math_mod.pow__float = math_mod.pow
math_mod.ceil__int = math_mod.ceil
math_mod.ceil__float = math_mod.ceil
math_mod.floor__int = math_mod.floor
math_mod.floor__float = math_mod.floor

return math_mod

File: lib/std/math.py
"""Standard library math module for Sapphire (Python target)."""

import math as _math

def abs(x):
  return -x if x < 0 else x

def sqrt(x):
  if x < 0:
    return None
  return _math.sqrt(x)

def min(a, b):
  return a if a < b else b

def max(a, b):
  return a if a > b else b

def safe_div(a, b):
  if b == 0 or b == 0.0:
    return None
  if isinstance(a, int) and isinstance(b, int):
    return a // b
  return a / b

def log(x, base=None):
  if x <= 0:
    return None
  if base is not None:
    if base <= 0 or base == 1:
      return None
    return _math.log(x, base)
  return _math.log(x)

def pow(base, exp):
  if isinstance(base, int) and isinstance(exp, int) and exp >= 0:
    return base ** exp
  return _math.pow(base, exp)

def ceil(x):
  return _math.ceil(x)

def floor(x):
  return _math.floor(x)

# Generic monomorphization aliases
abs__int = abs
abs__float = abs
sqrt__int = sqrt
sqrt__float = sqrt
min__int = min
min__float = min
max__int = max
max__float = max
safe_div__int = safe_div
safe_div__float = safe_div
log__int = log
log__float = log
pow__int = pow
pow__float = pow
ceil__int = ceil
ceil__float = ceil
floor__int = floor
floor__float = floor
